Dynamic healthchecks

winstonallo opened the pull request:
Closes: #8
Closes: #30

Add:
* New optional `healthcheck` field in the config, allows configuration of a command to run for healthchecks instead of waiting `starttime` to determine whether a process is healthy
* ~~New state `Starting` to differentiate between dynamic and simple healthchecks~~
* HealthCheckRunner struct to encapsulate the more complex healthcheck logic

HealthChecks are run in a tokio task with time out, their results are communicated via one-shot channels.

winstonallo committed:
Add basic structure for healthcheck events

winstonallo committed:
Run formatter

winstonallo committed:
Update spellcheck.yml

winstonallo committed:
Add spelling mistake

winstonallo committed:
Add name in log to verify spellcheck

winstonallo committed:
Add lowercase name

winstonallo committed:
Remove spellcheck tests

winstonallo committed:
Remove unused imports

winstonallo committed:
Add  struct and  state

winstonallo committed:
Add  state to differentiate from

winstonallo committed:
Remove nightly features from .rustfmt.toml

winstonallo committed:
Add command-based healthcheck

winstonallo committed:
Export healthcheck logic to different struct

winstonallo committed:
Simplify monitor_healthcheck, improve logs

winstonallo committed:
Move failure incrementation logic to failed_* functions

winstonallo committed:
Add comments to healthcheck config

winstonallo committed:
Fix log messages on process exit

winstonallo committed:
Merge branch 'main' into 8-dynamic-healtchecks

winstonallo committed:
Cache Instant::now() in ProcessState::fmt

winstonallo committed:
Add early return in monitor_healthcheck

fbruggem commented on src/run/proc.rs:
```diff
@@ -24,21 +24,24 @@ pub struct Process {
 impl Process {
     pub fn from_process_config(conf: conf::proc::ProcessConfig, proc_name: &str) -> Self {
         let is_autostart = conf.autostart();
+        let healthcheck = conf.healthcheck();
```
NIT PICK: cloning here and not on `conf: conf.clone()` should be less expensive because you only clone a subset of the struct and not all of it.

winstonallo committed:
Call clone on healthcheck rather than whole conf struct

winstonallo committed:
Compliance to our lord and savior clippy

fbruggem reviewed (changes requested):
LGTM - nicely done with the OneShotChannel!

TL;DR I'm fine with 2 States if we rename them to clearer names but still wanted to share one perspective
`Starting` =>`HealthCheckTime`
`HealthCheck` => `HealthCheckDynamic`

### Perspective
About the `Starting` state. I understand that having both in one makes the function more complex - this problem can be "solved/dirty fixed" by 

```rust
match p.is_dynamic_health() {
    Some(_) => dynamic(),
    None => timed(),
}
```

I would rather have a monitor function branched out with this match than another state which basically has the same "outside" behaviour. 
Because `Starting` and `HealthCheck` have the same communication to the "outside".
Either - failed try again until max tries
Or - succeeded NICE go into healthy

Also when we change something about how health-check works we might get code duplication or forget to change it in both places.

winstonallo commented:
> LGTM - nicely done with the OneShotChannel!
> 
> 
> 
> TL;DR I'm fine with 2 States if we rename them to clearer names but still wanted to share one perspective
> 
> `Starting` =>`HealthCheckTime`
> 
> `HealthCheck` => `HealthCheckDynamic`
> 
> 
> 
> ### Perspective
> 
> About the `Starting` state. I understand that having both in one makes the function more complex - this problem can be "solved/dirty fixed" by 
> 
> 
> 
> ```rust
> 
> match p.is_dynamic_health() {
> 
>     Some(_) => dynamic(),
> 
>     None => timed(),
> 
> }
> 
> ```
> 
> 
> 
> I would rather have a monitor function branched out with this match than another state which basically has the same "outside" behaviour. 
> 
> Because `Starting` and `HealthCheck` have the same communication to the "outside".
> 
> Either - failed try again until max tries
> 
> Or - succeeded NICE go into healthy
> 
> 
> 
> Also when we change something about how health-check works we might get code duplication or forget to change it in both places.

Sure we can do that, just having a hard time accepting starttime as a **health**check, a process stuck in an infinite loop would pass it. It's probably just me being pedantic though, I'll leave it as one state for now.

winstonallo committed:
Remove Starting state

winstonallo committed:
Merge branch '8-dynamic-healtchecks' of github.com:winstonallo/taskmaster into 8-dynamic-healtchecks

winstonallo committed:
Add fix for #30

winstonallo commented:
@FelixBrgm fixed 2 other things:
1. when a dynamic healthcheck fails, the process was always being restarted even if it was still running
2. fixed #30 - entries in the process table are kept until the parent calls wait() on the process, fixed by adding `let _ = p.exited();` in `monitor_stopped`

winstonallo committed:
Move starttime to healthcheck field of config, use enum in HealthCheck type for more robust deserialization

winstonallo committed:
Remove startup failures field from Process

winstonallo committed:
Fix failing test

winstonallo committed:
Change workingdir of test config to /tmp to avoid conflicts in Actions runners

winstonallo commented:
Also moved the entire healthcheck config to the HealthCheckRunner struct

winstonallo committed:
Remove commented out config

winstonallo committed:
Fix typo

winstonallo committed:
Fix further typo

winstonallo committed:
Reset healthcheck state when process exits in the middle of it

winstonallo commented:
merged by accident -.-

FelixBrgm commented:
Just checking - you marked this PR as a draft - so you will still change stuff? Because I still have the "Create review button"

fbruggem commented:
If its now one state it should be fine^*
